tools/clang/test/CodeGenDXIL/hlsl/intrinsics/buffer-agg-load-stores.hlsl

```diff
@@ -0,0 +1,231 @@
+// RUN: %dxc -DTYPE=uint64_t -DNUM=2 -DIX=SIx -T vs_6_6 %s | FileCheck %s
+// RUN: %dxc -DTYPE=double   -DNUM=2 -DIX=SIx -T vs_6_6 %s | FileCheck %s
+
+// RUN: %dxc -DOFF -DTYPE=float    -DNUM=4 -DIX=SIx -T vs_6_6 %s | FileCheck %s --check-prefixes=CHECK,OFF
+// RUN: %dxc -DOFF -DTYPE=bool     -DNUM=4 -DIX=SIx -T vs_6_6 %s | FileCheck %s --check-prefixes=CHECK,OFF
+// RUN: %dxc -DOFF -DTYPE=uint64_t -DNUM=2 -DIX=SIx -T vs_6_6 %s | FileCheck %s --check-prefixes=CHECK,OFF
+// RUN: %dxc -DOFF -DTYPE=double   -DNUM=2 -DIX=SIx -T vs_6_6 %s | FileCheck %s --check-prefixes=CHECK,OFF
+
+// RUN: %dxc -DARR -DTYPE=float    -DNUM=4 -DIX=VIx -T vs_6_6 %s | FileCheck %s --check-prefixes=CHECK,VIX
+// RUN: %dxc -DARR -DTYPE=bool     -DNUM=4 -DIX=VIx -T vs_6_6 %s | FileCheck %s --check-prefixes=CHECK,VIX
+// RUN: %dxc -DARR -DTYPE=uint64_t -DNUM=2 -DIX=VIx -T vs_6_6 %s | FileCheck %s --check-prefixes=CHECK,VIX
+// RUN: %dxc -DARR -DTYPE=double   -DNUM=2 -DIX=VIx -T vs_6_6 %s | FileCheck %s --check-prefixes=CHECK,VIX
+
+// RUN: %dxc -DTYPE=float    -DNUM=4 -DIX=VIx -T vs_6_6 %s | FileCheck %s --check-prefixes=CHECK,VIX
+// RUN: %dxc -DTYPE=bool     -DNUM=4 -DIX=VIx -T vs_6_6 %s | FileCheck %s --check-prefixes=CHECK,VIX
+// RUN: %dxc -DTYPE=uint64_t -DNUM=2 -DIX=VIx -T vs_6_6 %s | FileCheck %s --check-prefixes=CHECK,VIX
+// RUN: %dxc -DTYPE=double   -DNUM=2 -DIX=VIx -T vs_6_6 %s | FileCheck %s --check-prefixes=CHECK,VIX
+
+///////////////////////////////////////////////////////////////////////
+// Test codegen for various load and store operations and conversions
+//  for different aggregate buffer types and indices.
+///////////////////////////////////////////////////////////////////////
+
+
+
+// CHECK: %dx.types.ResRet.[[TY:[a-z][0-9][0-9]]] = type { [[TYPE:[a-z0-9]*]],
+
+#if defined(ARR)
+#define TYNAME(T,N) T[N]
+#define DECL(T,N,V) T V[N]
+#elif defined(MAT)
+#define TYNAME(T,N) matrix< T, N, N >
+#define DECL(T,N,V) matrix< T, N, N > V
+#elif defined(OFF)
+#define TYNAME(T,N) OffVector< T, N >
+#define DECL(T,N,V) OffVector< T, N > V
+#else
+#define TYNAME(T,N) Vector< T, N >
+#define DECL(T,N,V) Vector< T, N > V
+#endif
+
+template<typename T, int N>
+struct Vector {
+  vector<T, N> v;
+  Vector operator+(Vector vec) {
+    Vector ret;
+    ret.v = v + vec.v;
+    return ret;
+  }
+};
+
+template<typename T, int N>
+struct OffVector {
+  float4 pad1;
+  double pad2;
+  vector<T, N> v;
+  OffVector operator+(OffVector vec) {
+    OffVector ret;
+    ret.pad1 = 0.0;
+    ret.pad2 = 0.0;
+    ret.v = v + vec.v;
+    return ret;
+  }
+};
+
+  ByteAddressBuffer RoByBuf : register(t1);
+RWByteAddressBuffer RwByBuf : register(u1);
+
+  StructuredBuffer< TYNAME(TYPE,NUM) > RoStBuf : register(t2);
+RWStructuredBuffer< TYNAME(TYPE,NUM) > RwStBuf : register(u2);
+
+ConsumeStructuredBuffer< TYNAME(TYPE,NUM) > CnStBuf : register(u4);
+AppendStructuredBuffer< TYNAME(TYPE,NUM) > ApStBuf  : register(u5);
```

Comment:
Each of these buffer types take slightly or significantly different cod paths when generating loads and stores.